File: sdk/e2e/rust/tests/support/example_snapshot.rs
use cyt_indexer::build_catalog_index;
use serde_json::{json, Value};
use std::fs;
use std::io::{self, Write};
use std::path::{Path, PathBuf};

fn repo_root() -> PathBuf {
    Path::new(env!("CARGO_MANIFEST_DIR"))
        .join("../../..")
        .canonicalize()
        .expect("repo root")
}

pub fn parse_test_args() -> (Option<String>, Option<String>) {
    let file = std::env::var("CYT_E2E_FILE")
        .ok()
        .filter(|s| !s.is_empty())
        .or_else(|| std::env::var("FILE").ok().filter(|s| !s.is_empty()));
    let output = std::env::var("CYT_E2E_OUTPUT")
        .ok()
        .filter(|s| !s.is_empty())
        .or_else(|| std::env::var("OUTPUT").ok().filter(|s| !s.is_empty()));
    if file.is_some() || output.is_some() {
        return (file, output);
    }

    let args: Vec<String> = std::env::args().collect();
    let mut file = None;
    let mut output = None;
    let mut i = 0;
    while i < args.len() {
        match args[i].as_str() {
            "--file" if i + 1 < args.len() => {
                i += 1;
                file = Some(args[i].clone());
            }
            s if let Some(path) = s.strip_prefix("--file=") => {
                file = Some(path.to_string());
            }
            "--output" if i + 1 < args.len() => {
                i += 1;
                output = Some(args[i].clone());
            }
            s if let Some(path) = s.strip_prefix("--output=") => {
                output = Some(path.to_string());
            }
            _ => {}
        }
        i += 1;
    }
    (file, output)
}

pub fn resolve_snapshot_path(path: &str) -> PathBuf {
    let candidate = PathBuf::from(path);
    if candidate.is_file() {
        return candidate;
    }
    let from_repo = repo_root().join(path);
    if from_repo.is_file() {
        return from_repo;
    }
    panic!("snapshot file not found: {path} (also tried {})", from_repo.display());
}

pub fn load_snapshot(path: &Path) -> Value {
    let raw = fs::read_to_string(path).unwrap_or_else(|e| panic!("read {}: {e}", path.display()));
    serde_json::from_str(&raw).unwrap_or_else(|e| panic!("parse {}: {e}", path.display()))
}

fn enums_from_md(md_entries: &[Value]) -> Vec<Value> {
    md_entries
        .iter()
        .filter_map(|entry| entry.get("content").and_then(|v| v.as_str()))
        .map(|s| json!(s))
        .collect()
}

fn survivor_catalog(stage: &Value) -> Value {
    let mut survivor = json!({});
    if let Some(json_entries) = stage.get("json") {
        survivor["json"] = json_entries.clone();
    }
    if let Some(md_entries) = stage.get("md") {
        survivor["md"] = md_entries.clone();
    }
    survivor
}

pub fn extract_snapshot_parts(data: &Value) -> (Vec<Value>, Value, Vec<Value>) {
    let pruning = data.get("pruning").unwrap_or(&Value::Null);
    let stages = pruning
        .get("decomposed_catalog")
        .unwrap_or(&Value::Null);

    let (expected, build_stage, survivor_stage) = if data.get("body").is_some() {
        let expected = data["body"]
            .get("tools")
            .and_then(|v| v.as_array())
            .cloned()
            .unwrap_or_default();
        let build_stage = stages
            .get("build_index")
            .cloned()
            .unwrap_or(Value::Null);
        let survivor_stage = stages
            .get("rerank")
            .cloned()
            .unwrap_or_else(|| build_stage.clone());
        (expected, build_stage, survivor_stage)
    } else {
        let expected = data
            .get("tools")
            .and_then(|v| v.as_array())
            .cloned()
            .unwrap_or_default();
        let build_stage = stages
            .get("build_index")
            .cloned()
            .unwrap_or(Value::Null);
        let has_survivor = stages.get("json").is_some() || stages.get("md").is_some();
        let survivor_stage = if has_survivor {
            stages.clone()
        } else {
            build_stage.clone()
        };
        (expected, build_stage, survivor_stage)
    };

    let build_tools = build_stage
        .get("tools")
        .and_then(|v| v.as_array())
        .cloned()
        .unwrap_or_default();
    if build_tools.is_empty() && !expected.is_empty() {
        panic!(
            "snapshot has no pruning.decomposed_catalog.build_index.tools; cannot rebuild catalog index"
        );
    }

    let survivor = survivor_catalog(&survivor_stage);
    let has_json = survivor
        .get("json")
        .and_then(|v| v.as_array())
        .is_some_and(|a| !a.is_empty());
    let has_md = survivor
        .get("md")
        .and_then(|v| v.as_array())
        .is_some_and(|a| !a.is_empty());
    if !has_json && !has_md {
        panic!("snapshot has no rerank json/md entries for decomposition");
    }

    (build_tools, survivor, expected)
}

pub fn catalog_dict_from_snapshot(data: &Value) -> Value {
    let (build_tools, _survivor, _expected) = extract_snapshot_parts(data);
    let build_stage = data
        .pointer("/pruning/decomposed_catalog/build_index")
        .unwrap_or(&Value::Null);
    let md_entries = build_stage
        .get("md")
        .and_then(|v| v.as_array())
        .map(|a| a.as_slice())
        .unwrap_or(&[]);
    let enums = enums_from_md(md_entries);
    let index = build_catalog_index(&build_tools, &enums);
    index.to_catalog_dict("src/catalog")
}

pub fn write_output(catalog: &Value, output_path: Option<&str>) -> io::Result<()> {
    let payload = serde_json::to_string_pretty(catalog)? + "\n";
    match output_path {
        Some(path) => {
            let out = Path::new(path);
            if let Some(parent) = out.parent() {
                if !parent.as_os_str().is_empty() {
                    fs::create_dir_all(parent)?;
                }
            }
            fs::write(out, payload)
        }
        None => {
            io::stdout().write_all(payload.as_bytes())?;
            Ok(())
        }
    }
}

File: sdk/e2e/go/example_snapshot.go
package e2esupport

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"

	cytindexer "github.com/qdrddr/clear-your-tools/sdk/go"
)

func repoRoot() string {
	root, err := filepath.Abs(filepath.Join("..", "..", ".."))
	if err != nil {
		panic(err)
	}
	return root
}

func ParseTestArgs() (file *string, output *string) {
	if v := strings.TrimSpace(os.Getenv("CYT_E2E_FILE")); v != "" {
		file = &v
	} else if v := strings.TrimSpace(os.Getenv("FILE")); v != "" {
		file = &v
	}
	if v := strings.TrimSpace(os.Getenv("CYT_E2E_OUTPUT")); v != "" {
		output = &v
	} else if v := strings.TrimSpace(os.Getenv("OUTPUT")); v != "" {
		output = &v
	}
	if file != nil || output != nil {
		return file, output
	}

	args := os.Args
	for i := 0; i < len(args); i++ {
		switch {
		case args[i] == "--file" && i+1 < len(args):
			i++
			v := args[i]
			file = &v
		case strings.HasPrefix(args[i], "--file="):
			v := strings.TrimPrefix(args[i], "--file=")
			file = &v
		case args[i] == "--output" && i+1 < len(args):
			i++
			v := args[i]
			output = &v
		case strings.HasPrefix(args[i], "--output="):
			v := strings.TrimPrefix(args[i], "--output=")
			output = &v
		}
	}
	return file, output
}

func ResolveSnapshotPath(path string) string {
	if _, err := os.Stat(path); err == nil {
		return path
	}
	fromRepo := filepath.Join(repoRoot(), path)
	if _, err := os.Stat(fromRepo); err == nil {
		return fromRepo
	}
	panic("snapshot file not found: " + path + " (also tried " + fromRepo + ")")
}

func LoadSnapshot(path string) map[string]any {
	raw, err := os.ReadFile(path)
	if err != nil {
		panic("read " + path + ": " + err.Error())
	}
	var data map[string]any
	if err := json.Unmarshal(raw, &data); err != nil {
		panic("parse " + path + ": " + err.Error())
	}
	return data
}

func enumsFromMD(mdEntries []any) []string {
	enums := make([]string, 0, len(mdEntries))
	for _, entry := range mdEntries {
		obj, ok := entry.(map[string]any)
		if !ok {
			continue
		}
		content, ok := obj["content"].(string)
		if ok {
			enums = append(enums, content)
		}
	}
	return enums
}

func survivorCatalog(stage map[string]any) map[string]any {
	survivor := map[string]any{}
	if v, ok := stage["json"]; ok {
		survivor["json"] = v
	}
	if v, ok := stage["md"]; ok {
		survivor["md"] = v
	}
	return survivor
}

func asObject(v any) map[string]any {
	obj, _ := v.(map[string]any)
	if obj == nil {
		return map[string]any{}
	}
	return obj
}

func asArray(v any) []any {
	arr, _ := v.([]any)
	if arr == nil {
		return []any{}
	}
	return arr
}

func ExtractSnapshotParts(data map[string]any) ([]any, map[string]any, []any) {
	pruning := asObject(data["pruning"])
	stages := asObject(pruning["decomposed_catalog"])

	var expected, buildStage, survivorStage map[string]any
	if _, ok := data["body"]; ok {
		body := asObject(data["body"])
		expected = map[string]any{"tools": asArray(body["tools"])}
		buildStage = asObject(stages["build_index"])
		if v, ok := stages["rerank"]; ok {
			survivorStage = asObject(v)
		} else {
			survivorStage = buildStage
		}
	} else {
		expected = map[string]any{"tools": asArray(data["tools"])}
		buildStage = asObject(stages["build_index"])
		if stages["json"] != nil || stages["md"] != nil {
			survivorStage = stages
		} else {
			survivorStage = buildStage
		}
	}

	buildTools := asArray(buildStage["tools"])
	if len(buildTools) == 0 && len(asArray(expected["tools"])) > 0 {
		panic("snapshot has no pruning.decomposed_catalog.build_index.tools; cannot rebuild catalog index")
	}

	survivor := survivorCatalog(survivorStage)
	hasJSON := len(asArray(survivor["json"])) > 0
	hasMD := len(asArray(survivor["md"])) > 0
	if !hasJSON && !hasMD {
		panic("snapshot has no rerank json/md entries for decomposition")
	}

	return buildTools, survivor, asArray(expected["tools"])
}

func CatalogDictFromSnapshot(data map[string]any) (map[string]any, error) {
	buildTools, _, _ := ExtractSnapshotParts(data)
	pruning := asObject(data["pruning"])
	stages := asObject(pruning["decomposed_catalog"])
	buildStage := asObject(stages["build_index"])
	mdEntries := asArray(buildStage["md"])

	enumsJSON, err := json.Marshal(enumsFromMD(mdEntries))
	if err != nil {
		return nil, err
	}
	toolsJSON, err := json.Marshal(buildTools)
	if err != nil {
		return nil, err
	}
	indexJSON, err := cytindexer.BuildCatalogIndex(string(toolsJSON), string(enumsJSON))
	if err != nil {
		return nil, err
	}
	dictJSON, err := cytindexer.CatalogIndexToCatalogDict(indexJSON, "")
	if err != nil {
		return nil, err
	}
	var catalog map[string]any
	if err := json.Unmarshal([]byte(dictJSON), &catalog); err != nil {
		return nil, err
	}
	return catalog, nil
}

func WriteOutput(catalog map[string]any, outputPath *string) error {
	payload, err := json.MarshalIndent(catalog, "", "  ")
	if err != nil {
		return err
	}
	payload = append(payload, '\n')
	if outputPath == nil {
		_, err = os.Stdout.Write(payload)
		return err
	}
	if err := os.MkdirAll(filepath.Dir(*outputPath), 0o755); err != nil && !os.IsExist(err) {
		return err
	}
	return os.WriteFile(*outputPath, payload, 0o644)
}
